### ISSUE TITLE
Feature request: support replace-provider

### ISSUE THREAD

**Issue by mdb** [minamijoyo/tfmigrate#144]
@minamijoyo Would you consider a feature adding support for `terraform state replace-provider`? I'm imagining something like...

```hcl
migration "state" "update_provider" {
  dir = migrate.multi_state.mv_local_file_bar

  actions = [
    "replace-provider registry.terraform.io/-/datadog registry.terraform.io/datadog/datadog",
  ]
}
```

## More context

As for my use case, imagine a scenario like the following...

My team's `project-one` Terraform project uses Terraform 0.12.31. My team's `project-two` Terraform project uses Terraform 1.4.6. The CI/CD platform used by `project-one` and `project-two` utilizes [tfenv](https://github.com/tfutils/tfenv). As such, `tfenv` auto detects the presence of a `.terraform-version` file in `project-one` and in `project-two` and ensures the proper `terraform` CLI is used against each project. I want to migrate `datadog_dashboard.foo` from `project-one` to `project-two` using a `migration.hcl` file like the following:

```hcl
migration "multi_state" "mv_foo_dashboard" {
  from_dir = "../project-one"
  from_skip_plan = true

  to_dir   = "../project-two"

  actions = [
    "mv datadog_dashboard.foo datadog_dashboard.foo",
  ]
}
```

However, when attempting to use `tfmigrate` to migrate `datadog_dashboard.foo` from `project-one` to `project-two`, the following error is encountered, as `project-one`'s state uses an antiquated provider format that is not compatible with Terraform version 1.4.6 used by `project-two`:

```
tfmigrate plan migration.hcl
2023/08/08 13:19:10 [INFO] [runner] load migration file: migration.hcl
2023/08/08 13:19:10 [INFO] [migrator] multi start state migrator plan
2023/08/08 13:19:11 [INFO] [migrator@../project-one] terraform version: 0.12.31
2023/08/08 13:19:11 [INFO] [migrator@../project-one] initialize work dir
2023/08/08 13:20:40 [INFO] [migrator@../project-one] get the current remote state
2023/08/08 13:25:55 [INFO] [migrator@../project-one] override backend to local
2023/08/08 13:25:55 [INFO] [executor@../project-one] create an override file
2023/08/08 13:25:55 [INFO] [migrator@../project-one] creating local workspace folder in: ../project-one/terraform.tfstate.d/default
2023/08/08 13:25:55 [INFO] [executor@../project-one] switch backend to local
2023/08/08 13:26:58 [INFO] [migrator@../project-two] terraform version: 1.4.6
2023/08/08 13:26:58 [INFO] [migrator@../project-two] initialize work dir
2023/08/08 13:27:03 [INFO] [migrator@../project-two] get the current remote state
2023/08/08 13:27:05 [INFO] [migrator@../project-two] override backend to local
2023/08/08 13:27:05 [INFO] [executor@../project-two] create an override file
2023/08/08 13:27:05 [INFO] [migrator@../project-two] creating local workspace folder in: ../project-two/terraform.tfstate.d/default
2023/08/08 13:27:05 [INFO] [executor@../project-two] switch backend to local
2023/08/08 13:27:06 [INFO] [migrator] compute new states (../project-one => ../project-two)
2023/08/08 13:27:49 [INFO] [migrator@../project-one] skipping check diffs
2023/08/08 13:27:49 [INFO] [migrator@../project-two] check diffs
2023/08/08 13:27:51 [INFO] [executor@../project-two] remove the override file
2023/08/08 13:27:51 [INFO] [executor@../project-two] remove the workspace state folder
2023/08/08 13:27:51 [INFO] [executor@../project-two] switch back to remote
2023/08/08 13:27:54 [INFO] [executor@../project-one] remove the override file
2023/08/08 13:27:54 [INFO] [executor@../project-one] remove the workspace state folder
2023/08/08 13:27:54 [INFO] [executor@../project-one] switch back to remote
failed to run command (exited 1): terraform plan -state=/var/folders/46/sz1dzp417x7fk46kb3jv8cn00000gp/T/tmp4228698697 -out=/var/folders/46/sz1dzp417x7fk46kb3jv8cn00000gp/T/tfplan2353072760 -input=false -no-color -detailed-exitcode
stdout:

Warning: Version constraints inside provider configuration blocks are deprecated

  on provider.tf line 2, in provider "datadog":
   2:   version = "3.25.0"

Terraform 0.13 and earlier allowed provider version constraints inside the
provider configuration block, but that is now deprecated and will be removed
in a future version of Terraform. To silence this warning, move the provider
version constraint into the required_providers block.

stderr:

Error: failed to read schema for datadog_dashboard.foo in registry.terraform.io/-/datadog: failed to instantiate provider "registry.terraform.io/-/datadog" to obtain schema: unavailable provider "registry.terraform.io/-/datadog"
```

As a solution, could `tfmigrate` `replace-provider` support enable a `migration.hcl` like the following?

```hcl
migration "multi_state" "mv_foo_dashboard" {
  from_dir = "../project-one"
  from_skip_plan = true

  to_dir   = "../project-two"
  # plan won't work until after state replace-provider is performed, so we skip it
  to_skip_plan = true

  actions = [
    "mv datadog_dashboard.foo datadog_dashboard.foo",
  ]
}

migration "state" "update_provider" {
  dir = migrate.multi_state.mv_foo_dashboard.to_dir

  actions = [
    "replace-provider registry.terraform.io/-/datadog registry.terraform.io/datadog/datadog",
  ]
}
```

**Comment by minamijoyo** [comment]
Hi @mdb, The `replace-provider` action is a missing feature. It would be great if we had!

By the way, as I mentioned in #125, moving resources between different Terraform versions, especially from ancient before v0.14 to modern v1.x, is risky and may collapse your state. I recommend splitting resources anywhere once, following the upgrade guide to upgrade Terraform version and its state format, and then merging them into the final location.

As far as I can recall, there have been at least two significant changes to the state format between Terraform v0.12 and the current latest v1.5. The v0.13 introduced a namespace, and the v0.14 introduced a sensitive value. In general, v0.14 is forward compatible with v1.5, but I'm aware that some minor changes have been made, and I'm not sure that moving resources between different versions will not cause any problems.

**Comment by mdb** [comment]
> By the way, as I mentioned in https://github.com/minamijoyo/tfmigrate/issues/125, moving resources between different Terraform versions, especially from ancient before v0.14 to modern v1.x, is risky and may collapse your state

@minamijoyo Understood! Incidentally, this risk is part of the value I see in `tfmigrate`, which helps vet the safety of the resource move by performing a "dry run" migration and subsequent `terraform plan` against a local, dummy copy of the real state (assuming you haven't utilized the `to_skip_plan` feature I introduced in PR #143 , ironically ;) ).

> I recommend splitting resources anywhere once, following the upgrade guide to upgrade Terraform version and its state format

☝️ Just so I'm clear: Are you suggesting `tfmigrate`-ing resources to an intermediary state location, upgrading this intermediary state, then `tfmigrate`-ing resources from the intermediary state to the final state?

**Comment by minamijoyo** [comment]
Yes, I suggest completing the upgrade Terraform version steps outside of tfmigrate and using tfmigrate only for splitting and merging resources between directories.

According to the official Terraform upgrade guide, you need to terraform apply with v0.13 at least once before updating to v0.14. 
https://developer.hashicorp.com/terraform/language/v1.1.x/upgrade-guides/0-14
> Terraform v0.14 does not support legacy Terraform state snapshot formats from prior to Terraform v0.13, so before upgrading to Terraform v0.14 you must have successfully run terraform apply at least once with Terraform v0.13 so that it can complete its state format upgrades.

In addition, it is undefined how the terraform state mv command will behave if it reads the old state format. If you can see that the tfmigrate's multi_state mv migration can run the terraform plan command in to_dir with the new Terraform version, it indicates that it is at least able to read the old state format somehow, but I'm unsure this means the upgrade process was completed correctly or not.

If your strategy seems to work fine for you, it might be OK, at least in your case, but as the maintainer of tfmigrate, I have to say that it is unsafe in general.

**Comment by mdb** [comment]
> According to the official Terraform upgrade guide, you need to terraform apply with v0.13 at least once before updating to v0.14.
> ...
> In addition, it is undefined how the terraform state mv command will behave if it reads the old state format
> ...
> ...as the maintainer of tfmigrate, I have to say that it is unsafe in general.

Understood and agreed on all points! However, I do continue to think `tfmigrate` is uniquely beneficial in navigating such state operations, even if/when they are ill advised. Stay tuned; I'll try and cook up a `replace-provider` feature.

**Comment by minamijoyo** [comment]
Implemented it in v0.3.15 via #145
